### ISSUE TITLE
Switch to pwiz backend by default

### ISSUE THREAD

**Issue by sneumann** [sneumann/mzR#84]
Hi, mzR is still using the Ramp backend by default. XCMS has been calling 
the pwiz backend for at least half a year now. mzR should switch to backend pwiz 
in https://github.com/sneumann/mzR/blob/master/R/io.R#L1 by default. 
Should be done first thing in the next BioC-3.6 devel cycle in April/May 2017.

**Comment by lgatto** [comment]
There was a point where `pwiz` was default, but with erratic errors. Since then, many bugs have been fixed, so I would definitely for in favour of switching.

**Comment by lgatto** [comment]
Why wait until next release and not now?

**Comment by jotsetung** [comment]
I would also opt for now (in the devel branch) - then we have some time to test if all works correctly.

**Comment by jotsetung** [comment]
I'm hitting segfaults now with the `Ramp` backend when I'm extracting single spectra from an mzML file:
```r
extract chromatogram again ... Backend Ramp
[MSData::Spectrum::getMZIntensityPairs()] m/z array invalid size.

 *** caught segfault ***
address 0x0, cause 'memory not mapped'

Traceback:
 1: .External(list(name = "CppMethod__invoke_notvoid", address = <pointer: 0x7fda8840a660>,     dll = list(name = "Rcpp", path = "/Users/jo/R/2017-05/R-3.4-BioC3.5/lib/R/library/Rcpp/libs/x86_64/Rcpp.so",         dynamicLookup = TRUE, handle = <pointer: 0x7fda885555e0>,         info = <pointer: 0x7fda89055700>), numParameters = -1L),     <pointer: 0x7fda8862d350>, <pointer: 0x7fda8843eb60>, .pointer,     ...)
 2: object@backend$getPeakList(x)
 3: FUN(X[[i]], ...)
 4: lapply(X = X, FUN = FUN, ...)
 5: sapply(scans, function(x) object@backend$getPeakList(x)$peaks,     simplify = FALSE)
 6: sapply(scans, function(x) object@backend$getPeakList(x)$peaks,     simplify = FALSE)
 7: .peaks(object, scans)
 8: .local(object, ...)
 9: mzR::peaks(fileh, ms1fd$spIdx)
10: mzR::peaks(fileh, ms1fd$spIdx)
11: FUN(...)
12: doTryCatch(return(expr), name, parentenv, handler)
13: tryCatchOne(expr, names, parentenv, handlers[[1L]])
14: tryCatchList(expr, classes, parentenv, handlers)
15: tryCatch({    FUN(...)}, error = handle_error)
16: withCallingHandlers({    tryCatch({        FUN(...)    }, error = handle_error)}, warning = handle_warning)
17: FUN(X[[i]], ...)
18: lapply(X, FUN, ...)
19: bplapply(fDataPerFile, FUN = .applyFun2SpectraOfFileMulti, filenames = fNames,     queue = theQ, APPLYFUN = FUN, BPPARAM = BPPARAM, ...)
20: bplapply(fDataPerFile, FUN = .applyFun2SpectraOfFileMulti, filenames = fNames,     queue = theQ, APPLYFUN = FUN, BPPARAM = BPPARAM, ...)
21: .local(object, ...)
22: spectrapply(object, BPPARAM = BPPARAM)
23: spectrapply(object, BPPARAM = BPPARAM)
24: .local(object, ...)
25: spectra(cur_sample)
26: spectra(cur_sample)
27: (function (cur_sample, cur_file, rtm, mzm, aggFun) {    sps <- spectra(cur_sample)    rts <- rtime(cur_sample)    cur_res <- vector("list", nrow(rtm))    for (i in 1:nrow(rtm)) {        in_rt <- rts >= rtm[i, 1] & rts <= rtm[i, 2]        if (!any(in_rt)) {            cur_res[[i]] <- Chromatogram(filterMz = mzm[i, ],                 fromFile = as.integer(cur_file), aggregationFun = aggFun)            next        }        cur_sps <- lapply(sps[in_rt], function(spct, filter_mz,             aggFun) {            spct <- filterMz(spct, filter_mz)            if (!spct@peaksCount)                 return(c(NA_real_, NA_real_, missingValue))            return(c(range(spct@mz, na.rm = TRUE, finite = TRUE),                 do.call(aggFun, list(spct@intensity, na.rm = TRUE))))        }, filter_mz = mzm[i, ], aggFun = aggFun)        allVals <- unlist(cur_sps, use.names = FALSE)        idx <- seq(3, length(allVals), by = 3)        ints <- allVals[idx]        names(ints) <- names(cur_sps)        if (!all(is.na(ints))) {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = range(allVals[-idx], na.rm = TRUE,                   finite = TRUE), filterMz = mzm[i, ], fromFile = as.integer(cur_file),                 aggregationFun = aggFun)        }        else {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = mzm[i, ], filterMz = mzm[i,                   ], fromFile = as.integer(cur_file), aggregationFun = aggFun)        }    }    cur_res})(dots[[1L]][[1L]], dots[[2L]][[1L]], rtm = list(128, 148.6),     mzm = c(114.065617530061, 114.066758191939), aggFun = "sum")
28: .mapply(.FUN, dots, .MoreArgs)
29: FUN(...)
30: doTryCatch(return(expr), name, parentenv, handler)
31: tryCatchOne(expr, names, parentenv, handlers[[1L]])
32: tryCatchList(expr, classes, parentenv, handlers)
33: tryCatch({    FUN(...)}, error = handle_error)
34: withCallingHandlers({    tryCatch({        FUN(...)    }, error = handle_error)}, warning = handle_warning)
35: FUN(X[[i]], ...)
36: lapply(X, FUN, ...)
37: bplapply(X = seq_along(ddd[[1L]]), wrap, .FUN = FUN, .ddd = ddd,     .MoreArgs = MoreArgs, BPREDO = BPREDO, BPPARAM = BPPARAM)
38: bplapply(X = seq_along(ddd[[1L]]), wrap, .FUN = FUN, .ddd = ddd,     .MoreArgs = MoreArgs, BPREDO = BPREDO, BPPARAM = BPPARAM)
39: bpmapply(subs_by_file, seq_along(fileNames(subs)), FUN = function(cur_sample,     cur_file, rtm, mzm, aggFun) {    sps <- spectra(cur_sample)    rts <- rtime(cur_sample)    cur_res <- vector("list", nrow(rtm))    for (i in 1:nrow(rtm)) {        in_rt <- rts >= rtm[i, 1] & rts <= rtm[i, 2]        if (!any(in_rt)) {            cur_res[[i]] <- Chromatogram(filterMz = mzm[i, ],                 fromFile = as.integer(cur_file), aggregationFun = aggFun)            next        }        cur_sps <- lapply(sps[in_rt], function(spct, filter_mz,             aggFun) {            spct <- filterMz(spct, filter_mz)            if (!spct@peaksCount)                 return(c(NA_real_, NA_real_, missingValue))            return(c(range(spct@mz, na.rm = TRUE, finite = TRUE),                 do.call(aggFun, list(spct@intensity, na.rm = TRUE))))        }, filter_mz = mzm[i, ], aggFun = aggFun)        allVals <- unlist(cur_sps, use.names = FALSE)        idx <- seq(3, length(allVals), by = 3)        ints <- allVals[idx]        names(ints) <- names(cur_sps)        if (!all(is.na(ints))) {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = range(allVals[-idx], na.rm = TRUE,                   finite = TRUE), filterMz = mzm[i, ], fromFile = as.integer(cur_file),                 aggregationFun = aggFun)        }        else {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = mzm[i, ], filterMz = mzm[i,                   ], fromFile = as.integer(cur_file), aggregationFun = aggFun)        }    }    cur_res}, MoreArgs = list(rtm = rt, mzm = mz, aggFun = aggregationFun),     BPPARAM = BPPARAM, SIMPLIFY = FALSE)
40: bpmapply(subs_by_file, seq_along(fileNames(subs)), FUN = function(cur_sample,     cur_file, rtm, mzm, aggFun) {    sps <- spectra(cur_sample)    rts <- rtime(cur_sample)    cur_res <- vector("list", nrow(rtm))    for (i in 1:nrow(rtm)) {        in_rt <- rts >= rtm[i, 1] & rts <= rtm[i, 2]        if (!any(in_rt)) {            cur_res[[i]] <- Chromatogram(filterMz = mzm[i, ],                 fromFile = as.integer(cur_file), aggregationFun = aggFun)            next        }        cur_sps <- lapply(sps[in_rt], function(spct, filter_mz,             aggFun) {            spct <- filterMz(spct, filter_mz)            if (!spct@peaksCount)                 return(c(NA_real_, NA_real_, missingValue))            return(c(range(spct@mz, na.rm = TRUE, finite = TRUE),                 do.call(aggFun, list(spct@intensity, na.rm = TRUE))))        }, filter_mz = mzm[i, ], aggFun = aggFun)        allVals <- unlist(cur_sps, use.names = FALSE)        idx <- seq(3, length(allVals), by = 3)        ints <- allVals[idx]        names(ints) <- names(cur_sps)        if (!all(is.na(ints))) {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = range(allVals[-idx], na.rm = TRUE,                   finite = TRUE), filterMz = mzm[i, ], fromFile = as.integer(cur_file),                 aggregationFun = aggFun)        }        else {            cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                 intensity = ints, mz = mzm[i, ], filterMz = mzm[i,                   ], fromFile = as.integer(cur_file), aggregationFun = aggFun)        }    }    cur_res}, MoreArgs = list(rtm = rt, mzm = mz, aggFun = aggregationFun),     BPPARAM = BPPARAM, SIMPLIFY = FALSE)
41: withCallingHandlers(expr, warning = function(w) invokeRestart("muffleWarning"))
42: suppressWarnings(res <- bpmapply(subs_by_file, seq_along(fileNames(subs)),     FUN = function(cur_sample, cur_file, rtm, mzm, aggFun) {        sps <- spectra(cur_sample)        rts <- rtime(cur_sample)        cur_res <- vector("list", nrow(rtm))        for (i in 1:nrow(rtm)) {            in_rt <- rts >= rtm[i, 1] & rts <= rtm[i, 2]            if (!any(in_rt)) {                cur_res[[i]] <- Chromatogram(filterMz = mzm[i,                   ], fromFile = as.integer(cur_file), aggregationFun = aggFun)                next            }            cur_sps <- lapply(sps[in_rt], function(spct, filter_mz,                 aggFun) {                spct <- filterMz(spct, filter_mz)                if (!spct@peaksCount)                   return(c(NA_real_, NA_real_, missingValue))                return(c(range(spct@mz, na.rm = TRUE, finite = TRUE),                   do.call(aggFun, list(spct@intensity, na.rm = TRUE))))            }, filter_mz = mzm[i, ], aggFun = aggFun)            allVals <- unlist(cur_sps, use.names = FALSE)            idx <- seq(3, length(allVals), by = 3)            ints <- allVals[idx]            names(ints) <- names(cur_sps)            if (!all(is.na(ints))) {                cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                   intensity = ints, mz = range(allVals[-idx],                     na.rm = TRUE, finite = TRUE), filterMz = mzm[i,                     ], fromFile = as.integer(cur_file), aggregationFun = aggFun)            }            else {                cur_res[[i]] <- Chromatogram(rtime = rts[in_rt],                   intensity = ints, mz = mzm[i, ], filterMz = mzm[i,                     ], fromFile = as.integer(cur_file), aggregationFun = aggFun)            }        }        cur_res    }, MoreArgs = list(rtm = rt, mzm = mz, aggFun = aggregationFun),     BPPARAM = BPPARAM, SIMPLIFY = FALSE))
43: .extractMultipleChromatograms(object, rt = rt, mz = mz, aggregationFun = aggregationFun,     missingValue = missing)
44: .local(object, ...)
45: extractChromatograms(object, rt = rt, mz = mz, aggregationFun = aggregationFun,     missing = missing)
46: extractChromatograms(object, rt = rt, mz = mz, aggregationFun = aggregationFun,     missing = missing)
47: .local(object, ...)
48: extractChromatograms(cur_obj, mz = mzr_new, rt = rtr)
49: extractChromatograms(cur_obj, mz = mzr_new, rt = rtr)

Possible actions:
1: abort (with core dump, if enabled)
2: normal R exit
3: exit R without saving workspace
4: exit R saving workspace
Selection: 3

```

I'm now evaluating if this doesn't happen with the `pwiz` backend.

**Comment by jotsetung** [comment]
The segfault above does no longer occur if the `pwiz` backend is used - another argument to switch from `Ramp` to `pwiz`.

**Comment by lgatto** [comment]
@jotsetung I have merged the pull request and will push to Bioc. I'll leave this issue open until this change has been announced on the devel mailing list. Whoever announces it should then close this issue.
